File: 02-observer/weather-node/weatherdata/weatherdata.ts
import Observer from "../displays/observer";
import Subject from "./subject";

export default class WeatherData implements Subject {
  private observers: Array<Observer>;
  private temperature: number;
  private humidity: number;
  private pressure: number;

  constructor() {
    this.observers = [];
    this.temperature = 0;
    this.humidity = 0;
    this.pressure = 0;
  }

  public registerObservers(o: Observer): void {
    this.observers.push(o);
  }

  public removeObserver(o: Observer): void {
    const i = this.observers.indexOf(o);
    if (i >= 0) {
      this.observers.splice(i, 1);
    }
  }

  public notifyObservers(): void {
    for (const o of this.observers) {
      o.update(this.temperature, this.humidity, this.pressure);
    }
  }

  public measurementsChanges(): void {
    this.notifyObservers();
  }

  public setMeasurements(
    temperature: number,
    humidity: number,
    pressure: number
  ): void {
    this.temperature = temperature;
    this.humidity = humidity;
    this.pressure = pressure;
    this.measurementsChanges();
  }
}

File: 02-observer/weather-go/weatherdata/weatherdata.go
package weatherdata

import (
	"slices"

	"github.com/k-nox/design-patterns/02-observer/weather-go/observer"
)

type WeatherData struct {
	observers []observer.Observer
	temp float32
	humidity float32
	pressure float32
}

func New() *WeatherData {
	return &WeatherData{
		observers: []observer.Observer{},
	}
}

func (w *WeatherData) RegisterObserver(o observer.Observer) {
	w.observers = append(w.observers, o)
}

func (w *WeatherData) RemoveObserver(o observer.Observer) {
	i := slices.Index(w.observers, o)
	if i >= 0 {
		w.observers = slices.Delete(w.observers, i, i+1)
	}
}

func (w *WeatherData) NotifyObservers() {
	for _, o := range w.observers {
		o.Update(w.temp, w.humidity, w.pressure)
	}
}

func (w *WeatherData) measurementsChanged() {
	w.NotifyObservers()
}

func (w *WeatherData) SetMeasurements(temp float32, humidity float32, pressure float32) {
	w.temp = temp
	w.humidity = humidity
	w.pressure = pressure
	w.measurementsChanged()
}


